jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,517 @@
+
+        // Obtain a tracker, creating a new one if necessary.  Trackers are removed if CyclicTimeouts#onExpired returns true
+        Tracker tracker = _trackers.computeIfAbsent(id, this::newTracker);
+
+        // If we are not over-limit then handle normally
+        if (!tracker.isRateExceeded(request.getBeginNanoTime(), true))
+            return nextHandler(request, response, callback);
+
+        // Otherwise reject the request
+        return _rejectHandler.handle(request, response, callback);
+    }
+
+    Tracker newTracker(String id)
+    {
+        return new Tracker(id, _rateControlFactory.newRate());
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        _cyclicTimeouts = new CyclicTimeouts<>(getServer().getScheduler())
+        {
+            @Override
+            protected Iterator<Tracker> iterator()
+            {
+                return _trackers.values().iterator();
+            }
+
+            @Override
+            protected boolean onExpired(Tracker tracker)
+            {
+                return tracker.isIdle(System.nanoTime());
+            }
+        };
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        _cyclicTimeouts.destroy();
+        _cyclicTimeouts = null;
+    }
+
+    /**
+     * A RateTracker is associated with a connection, and stores request rate data.
+     */
+    class Tracker implements CyclicTimeouts.Expirable
+    {
+        private final AutoLock _lock = new AutoLock();
+        private final String _id;
+        private final RateControl _rateControl;
+        private long _expireAt;
+
+        Tracker(String id, RateControl rateControl)
+        {
+            _id = id;
+            _rateControl = rateControl;
+            _expireAt = System.nanoTime() + TimeUnit.SECONDS.toNanos(2);
+        }
+
+        public String getId()
+        {
+            return _id;
+        }
+
+        RateControl getRateControl()
+        {
+            return _rateControl;
+        }
+
+        public boolean isRateExceeded(long now, boolean addSample)
+        {
+            try (AutoLock l = _lock.lock())
+            {
+                CyclicTimeouts<Tracker> cyclicTimeouts = _cyclicTimeouts;
+                if (addSample && cyclicTimeouts != null)
+                {
+                    // schedule a check to remove this tracker if idle
+                    _expireAt = now + TimeUnit.SECONDS.toNanos(2L);
+                    cyclicTimeouts.schedule(this);
+                }
+                return _rateControl.isRateExceeded(now, addSample);
+            }
+        }
+
+        public boolean isIdle(long now)
+        {
+            try (AutoLock l = _lock.lock())
+            {
+                return _rateControl.isIdle(now);
+            }
+        }
+
+        @Override
+        public long getExpireNanoTime()
+        {
+            return _expireAt;
```

Comment:
This would return 0 if `isRateExceeded()` hasn't yet been called. Maybe `_expireAt` should default to `Long.MIN_VALUE`?

Reply:
See constructor.